---
title: "Genome Annotation and the Interpretation of Gene Lists"
author: "Kamron Kimball"
date: "2/27/2020"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(DESeq2)
metafile <- "GSE37704_metadata.csv"
countfile <- "GSE37704_featurecounts.csv"
```

The data for for hands-on session comes from GEO entry: GSE37704, which is associated with the following publication:

Trapnell C, Hendrickson DG, Sauvageau M, Goff L et al. "Differential analysis of gene regulation at transcript resolution with RNA-seq". Nat Biotechnol 2013 Jan;31(1):46-53. PMID: 23222703

Here we read the two input files

```{r}
# Import metadata and take a peak
colData = read.csv(metafile, row.names=1)
head(colData)

#And the count file
countData = read.csv(countfile, row.names=1)
head(countData)
```

There are `r nrow(countData)` genes in this dataset and `r ncol(countData)` experiments (samples)...

It looks like there is a funny extra 'length' column in our countData that we need to remove.
 
```{r}
# Note we need to remove the odd first $length col
countData <- as.matrix(countData[,-1])
head(countData)
```

We need to tidy up a little bit more, because we need to remove the genes where we have no data. (0 in all experiments)

Lets take the rowSums() function and exclude the rows with 0 values

```{r}
# Filter count data where you have 0 read count across all samples.
cleandata = countData[rowSums(countData) !=0,]
head(cleandata)
```

Check how many genes we have left

```{r}
nrow(cleandata)
```

```{r}
dds = DESeqDataSetFromMatrix(countData=cleandata,
                             colData=colData,
                             design=~condition)
dds = DESeq(dds)
```

Lets see our results

```{r}
res <- results(dds)
res
```

## Volcono plot
Summary volcono plot of our results to date.

```{r}
plot(res$log2FoldChange, -log(res$padj))
abline(h=-log(0.05), col="red")
abline(v=c(-2,2), col="red")
```

```{r}
library("AnnotationDbi")
library("org.Hs.eg.db")

columns(org.Hs.eg.db)
```

lets add extra annotations to our result object

```{r}
res$symbol = mapIds(org.Hs.eg.db,
                    keys=row.names(res),
                    keytype = "ENSEMBL",
                    column = "SYMBOL",
                    multiVals="first")

res$entrez <- mapIds(org.Hs.eg.db,
                    keys=row.names(res),
                    keytype="ENSEMBL",
                    column="ENTREZID",
                    multiVals="first")

head(res)
```

Now we have our geneIDs in the correct format we can do pathway analysis

## Pathway Analysis

Here we are going to use the gage package for pathway analysis. Once we have a list of enriched pathways, we're going to use the pathview package to draw pathway diagrams, shading the molecules in the pathway by their degree of up/down-regulation.

```{r}
library(pathview)
library(gage)
library(gageData)

data(kegg.sets.hs)
data(sigmet.idx.hs)

# Focus on signaling and metabolic pathways only
kegg.sets.hs = kegg.sets.hs[sigmet.idx.hs]

# Examine the first 3 pathways
head(kegg.sets.hs, 3)
```

Make a input vector of our numeric quantity of interest i.e. fold change values and make sure it has ENTREZ gene names associated with it

```{r}
foldchanges = res$log2FoldChange
names(foldchanges) = res$entrez
head(foldchanges)
```

```{r}
# Get the results
keggres = gage(foldchanges, gsets=kegg.sets.hs)
```

```{r}
attributes(keggres)

# Look at the first few down (less) pathways
head(keggres$less)
```

We use the KEGG pathway ID from our top hit along with our `foldchange` vector to view our results as a more useful figure.

```{r}
pathview(gene.data=foldchanges, pathway.id="hsa04110")
```

![hsa04110.pathview.png]()

We can also generate figures for other pathways or make different figures of the same pathways

```{r}
#A different PDF based output of the same data
pathview(gene.data=foldchanges, pathway.id="hsa04110", kegg.native=FALSE)

#A different output
pathview(gene.data=foldchanges, pathway.id="hsa03030")
```

Now, let's process our results a bit more to automagicaly pull out the top 5 upregulated pathways, then further process that just to get the pathway IDs needed by the pathview() function. We'll use these KEGG pathway IDs for pathview plotting below.

```{r}
## Focus on top 5 upregulated pathways here for demo purposes only
keggrespathways <- rownames(keggres$greater)[1:5]

# Extract the 8 character long IDs part of each string
keggresids = substr(keggrespathways, start=1, stop=8)
keggresids
```

Finally, lets pass these IDs in keggresids to the pathview() function to draw plots for all the top 5 pathways.

```{r}
pathview(gene.data=foldchanges, pathway.id=keggresids, species="hsa")
```

```{r}
knitr::include_graphics("hsa00140.pathview.png")
```

















